### PR TITLE
refactor(plexus): convert SvgDefEntry from class to functional component

### LINKED ISSUE #3395: [UI Refactor] [plexus] Migrate SvgDefEntry to Functional Component
This issue is part of the larger effort to modernize the Jaeger UI codebase (Parent Issue #2610). The goal is to refactor the `SvgDefEntry` component from a React Class Component to a React Functional Component using Hooks.

**Target Component:** `packages/plexus/src/Digraph/SvgDefEntry.tsx`

> **Note:** This component is part of the `plexus` package, which is a graph visualization library used by Jaeger UI.

### Acceptance Criteria

- [ ] The component is converted to a Functional Component.
- [ ] `this.state` is replaced with `useState` or `useReducer`.
- [ ] Lifecycle methods (componentDidMount, etc.) are replaced with `useEffect`.
- [ ] `memoizeOne` calls are converted to `useMemo` where appropriate.
- [ ] Static properties are handled appropriately (e.g., moved outside component or use module-level constants).
- [ ] All existing PropType validations are preserved.
- [ ] **Performance:** Wrap in `React.memo` to maintain `PureComponent` behavior.
- [ ] **Visual Verification:** Attach "Before" and "After" screenshots to the PR to prove no visual regression.
- [ ] **Tests:** Existing unit tests must pass.

### Plexus-Specific Guidance

| Pattern | Class Approach | Functional Approach |
| --- | --- | --- |
| **Instance ID** | `baseId = \`prefix--${idCounter++}\`` | `const baseId = useRef(\`prefix--${idCounter++}\`).current` |
| **Memoization** | `memoizeOne(fn)` | `useMemo(() => fn, [deps])` or keep `memoizeOne` with `useRef` |
| **Static Props** | `static defaultProps = {...}` | Function parameter defaults or separate constant |
| **Render Utils** | `this.renderUtils = {...}` | `const renderUtils = useMemo(() => ({...}), [deps])` |

### Migration Guide (Cheat Sheet)

| Feature | Old (Class) | New (Functional) |
| --- | --- | --- |
| **State** | `this.state = { isOpen: false }` | `const [isOpen, setIsOpen] = useState(false);` |
| **Side Effects** | `componentDidMount()` | `useEffect(() => { ... }, [])` |
| **Props** | `this.props.traceId` | `props.traceId` (or destructure: `const { traceId } = props;`) |
| **Default Props** | `static defaultProps = { name: 'Guest' }` | `function Component({ name = 'Guest' }) { ... }` |
| **Performance** | `class MyComp extends React.PureComponent` | `export default React.memo(MyComp)` |
| **Context** | `static contextType = MyContext` | `const val = useContext(MyContext)` |
| **Refs** | `this.myRef = React.createRef()` | `const myRef = useRef(null)` |

### Common Pitfalls (Read Carefully!)

- **Stale Closures:** Be careful when using `useEffect`. If you use a variable inside the effect that comes from props or state, it must be in the dependency array. If you omit it, the effect will use the old value.
- **Testing with Enzyme:** Functional components do not have instances. Tests like `wrapper.instance().method()` or `wrapper.state('value')` will fail. You should refactor these tests to simulate events (clicks) and check the rendered output instead.
- **Static Properties:** Class components can have static properties attached directly. For functional components, you'll need to assign these after the function definition or use module-level constants.

### Resources

- [React Official Guide: Hooks at a Glance](https://reactjs.org/docs/hooks-overview.html)
- [A Complete Guide to useEffect (Overreacted.io)](https://overreacted.io/a-complete-guide-to-useeffect/) - Essential reading for understanding dependency arrays.
- [React Redux Hooks Documentation](https://react-redux.js.org/api/hooks) - Guide for replacing `connect()`.
- [Thinking in React Hooks](https://wattenberger.com/blog/react-hooks) - Visual guide to the mental model shift.
- [How to fetch data with React Hooks](https://www.robinwieruch.de/react-hooks-fetch-data) - Handling async operations.

### PR DESCRIPTION
**Pull request by thc1006**
## Summary
- Convert the SvgDefEntry component from class-based to functional component
- Use React.memo for shallow comparison (equivalent to PureComponent)
- Add comprehensive test suite with 17 tests

## Changes
- Replace class component with arrow function component
- Preserve generic type parameters with trailing comma syntax
- Preserve renderDefaultMarker helper function unchanged

## Test Coverage
- Basic rendering and custom renderEntry support
- Global ID generation via renderUtils
- className application
- setOnEntry prop handling
- Default marker rendering with scaling
- renderEntry callback parameter passing
- React.memo verification

Closes #3395

Signed-off-by: thc1006 <84045975+thc1006@users.noreply.github.com>

### CODE REVIEW

**Comment by jkowall** [comment]
@thc1006 please review the CI errors and copilot feedback. Thanks!

File: packages/plexus/src/Digraph/SvgDefEntry.test.js [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        </svg>
+      );
+      expect(customRender.mock.calls[0][2]).toBe('global-entry-id');
+    });
+  });
+
+  describe('React.memo behavior', () => {
+    it('is wrapped with React.memo for performance', () => {
+      expect(SvgDefEntry.$$typeof).toBeDefined();
+    });
```
[the remark this answers is not in the document]
Comment (thc1006) [reply packages/plexus/src/Digraph/SvgDefEntry.test.js]: This suggestion has already been applied.

**Current code (line 276):**
\
**Technical clarification:** The statement that "\ property exists on all React elements" is not entirely accurate:

| Component Type | \ Value |
|----------------|---------------------|
| Regular function component | \ |
| \ | \ |
| \ | \ |

The \ property only exists on special React wrapper types (memo, forwardRef, lazy, etc.), not on plain function components. However, using the specific symbol check is indeed the better approach as it:
1. Ensures it's specifically \ (not accidentally \)
2. Provides clearer test intent

Test verified passing: 1 passed, 16 skipped (other tests filtered out).

File: packages/plexus/src/Digraph/SvgDefEntry.tsx [hunk 2]
```diff
@@ -47,16 +47,17 @@ function renderDefaultMarker(
...
+  const entryProps = assignMergeCss(
+    {
+      className: getClassName('DefEntry'),
+    },
+    getProps(setOnEntry, graphState)
+  );
+  return renderEntry(graphState, entryProps, id);
+};
+
+// React.memo provides shallow comparison equivalent to PureComponent
```
Comment (Parship12) [inline packages/plexus/src/Digraph/SvgDefEntry.tsx]: it explains the "what", not a non-obvious constraint. Please remove it.

Comment (ghost) [inline packages/plexus/src/Digraph/SvgDefEntry.tsx]: Redundant comment

File: packages/plexus/src/Digraph/SvgDefEntry.test.js [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+            <SvgDefEntry {...defaultProps} graphState={graphState2} />
+          </defs>
+        </svg>
+      );
+      const marker2 = c2.querySelector('marker');
+      const height2 = marker2.getAttribute('markerHeight');
+
+      // Both should have some height value (scaling logic is mocked)
+      expect(height1).toBeTruthy();
+      expect(height2).toBeTruthy();
```
Comment (ghost) [inline packages/plexus/src/Digraph/SvgDefEntry.test.js:234]: `getValueScaler: () => () => 0.5,`
As `getValueScaler` is mocked to return 0.5, so these heights will be identical. We need to un-mock `getValueScaler` (or use the real implementation) to actually test that.

File: packages/plexus/src/Digraph/SvgDefEntry.test.js [hunk 4]
```diff
@@ -0,0 +1,298 @@
+// Copyright (c) 2026 Uber Technologies, Inc.
```
Comment (yurishkuro) [inline packages/plexus/src/Digraph/SvgDefEntry.test.js]:
```suggestion
// Copyright (c) 2026 The Jaeger Authors
```

**Comment by yurishkuro** [comment]
what is the story with two open comments?

**Comment by thc1006** [comment]
Dear @yurishkuro ,
Sorry, should have resolved these earlier. Both are addressed in the code.

Parship12's comment about getValueScaler being mocked to a constant was fixed in e3928c22, switching to actual.getValueScaler so the scaling test exercises real logic. (Parship12 approved after the fix.)

The Copilot TDZ suggestion is a false positive. mockAssignMergeCssResults and mockGetPropsResults are only read during test execution (component render path), not at module load time, so no TDZ issue. All 17 tests pass.

I'll resolve both threads. Let me know if you'd rather I apply the Copilot suggestion anyway.

**Comment by thc1006** [comment]
Just noticed that SvgDefEntry.tsx was already converted in #3528 by @hharshhsaini. Closing this since the conversion is already on main.

@Parship12 thank you for taking the time to review and approve this. The feedback on the getValueScaler mock and the redundant comment both made the tests better, and I'll carry those lessons forward.

Similar to #3402, this branch has a 298-line test file for SvgDefEntry covering rendering, gradient calculations, and edge cases. There don't seem to be tests for this component on main yet.

If a standalone test PR (migrated to Vitest) would be useful, I'm happy to submit one. Let me know.
